internal/app/db.go

```diff
@@ -0,0 +1,330 @@
+			if column.IsPrimary {
+				pks = column.Name
+			}
+		} else {
+			cols = fmt.Sprintf("%s,%s", cols, col)
+			if column.IsPrimary {
+				pks = fmt.Sprintf("%s,%s", pks, column.Name)
+			}
+		}
+	}
+
+	if pks != "" {
+		cols = fmt.Sprintf("%s,PRIMARY KEY (%s)", cols, pks)
+	}
+
+	if cols == "" {
+		return "", errors.New("schema must have at least one column")
+	}
+
+	stmt := fmt.Sprintf(
+		"CREATE TABLE IF NOT EXISTS %s (%s)",
+		dbm.table, cols)
+	return stmt, nil
+}
+
+// Export exports the current db to a parquet file at the given path.
+func (dbm *DBManager) Export(ctx context.Context, exportPath string) error {
+	var err error
+	db := dbm.db
+	// db is nil before replication starts.
+	// In that case, we open all existing db files
+	// and upload them.
+	if db == nil {
+		// convert the export path to a db path:
+		// <ts>.db.parquet -> <ts>.db
+		dbPath := strings.ReplaceAll(exportPath, ".parquet", "")
+		db, err = sql.Open("duckdb", dbPath)
+		if err != nil {
+			return err
+		}
+		defer func() {
+			if err := db.Close(); err != nil {
+				fmt.Printf("cannot close db %v \n", err)
+			}
+		}()
+		slog.Info("backing up db", "at", exportPath)
+	} else {
+		slog.Info("backing up current db")
+	}
+	_, err = db.ExecContext(ctx,
+		fmt.Sprintf(
+			`INSTALL parquet;
+			 LOAD parquet;
+			 COPY (SELECT * FROM %s) TO '%s' (FORMAT PARQUET)`,
+			dbm.table, exportPath))
+	if err != nil {
+		return fmt.Errorf("cannot export to parquet file: %s", err)
+	}
+
+	return nil
+}
+
+func (dbm *DBManager) cleanup(dbPath string) error {
+	slog.Info("deleting db dump", "at", dbPath)
+	if err := os.Remove(dbPath); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot delete file: %s", err)
+		}
+	}
+
+	walPath := dbPath + ".wal"
+	slog.Info("deleting db wal", "at", walPath)
+	if err := os.Remove(walPath); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot delete file: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// UploadAt uploads a db dump at the given path.
+// It returns an error if a dumps cannot be uploaded.
+// It cleans up the db dumps and export files after uploading.
+func (dbm *DBManager) UploadAt(ctx context.Context, exportPath string) error {
+	f, err := os.Open(exportPath)
+	if err != nil {
+		return fmt.Errorf("cannot open file: %s", err)
+	}
+
+	fi, err := f.Stat()
+	if err != nil {
+		return fmt.Errorf("cannot stat file: %s", err)
+	}
+
+	progress := progressbar.DefaultBytes(
+		fi.Size(),
+		"Uploading file...",
+	)
+
+	if err := dbm.uploader.Upload(ctx, exportPath, progress); err != nil {
+		return fmt.Errorf("upload: %s", err)
+	}
+
+	// cleanup the exported parquet file
+	slog.Info("deleting db parquet export", "at", exportPath)
+	if err := os.Remove(exportPath); err != nil {
+		if !os.IsNotExist(err) {
+			return fmt.Errorf("cannot delete file: %s", err)
+		}
+	}
+
+	return nil
+}
+
+// UploadAll uploads all db dumps in the db dir.
+func (dbm *DBManager) UploadAll(ctx context.Context) error {
```

Comment:
`UploadAll` is called when the replication starts to send any remaining data left.